File: backend/domain/layout.py
import re
from dataclasses import dataclass

from .fonts.base import Font, Glyph

SCREEN_WIDTH = 256
SCREEN_HEIGHT = 192


@dataclass(frozen=True, slots=True)
class ScreenConfig:
    x_pos: int
    y_pos: int
    new_line: int
    width: int = SCREEN_WIDTH
    height: int = SCREEN_HEIGHT
    matches: tuple[str, ...] = ()
    tags: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class PlacedGlyph:
    codepoint: int
    x: int
    y: int
    glyph: Glyph


@dataclass(frozen=True, slots=True)
class LayoutResult:
    glyphs: tuple[PlacedGlyph, ...]
    overflow_horizontal: bool = False
    overflow_vertical: bool = False


def _strip_tags(line: str, tags: tuple[str, ...]) -> str:
    for tag in tags:
        line = re.sub(tag, "", line)
    return line


def _pivot_line(lines: list[str], matches: tuple[str, ...], cursor_line: int) -> int:
    index = min(cursor_line, len(lines) - 1)
    while index >= 0:
        if any(re.match(pattern, lines[index]) for pattern in matches):
            return index + 1
        index -= 1
    return 0


def layout(text: str, font: Font, config: ScreenConfig, cursor_line: int = 0) -> LayoutResult:
    lines = text.split("\n")
    start = _pivot_line(lines, config.matches, cursor_line) if config.matches else 0

    placed: list[PlacedGlyph] = []
    overflow_horizontal = False
    overflow_vertical = False
    y = config.y_pos

    for line in lines[start:]:
        if y > config.height:
            overflow_vertical = True
            break

        x = config.x_pos
        for char in _strip_tags(line, config.tags):
            codepoint = ord(char)
            glyph = font.glyph(codepoint)
            if glyph is None:
                continue
            placed.append(PlacedGlyph(codepoint, x + glyph.leading, y, glyph))
            x += glyph.advance + 1
            if x > config.width:
                overflow_horizontal = True
                break
        y += config.new_line

    return LayoutResult(tuple(placed), overflow_horizontal, overflow_vertical)

File: frontend/src/layout.ts
import type { AtlasMetrics, LayoutGlyph, Screen } from "./api";

export const SCREEN_WIDTH = 256;
export const SCREEN_HEIGHT = 192;

export interface ClientLayoutResult {
  glyphs: LayoutGlyph[];
  overflowHorizontal: boolean;
  overflowVertical: boolean;
}

function stripTags(line: string, tags: string[]): string {
  let result = line;
  for (const tag of tags) {
    result = result.replace(new RegExp(tag, "g"), "");
  }
  return result;
}

function matchesAtStart(line: string, patterns: string[]): boolean {
  return patterns.some((pattern) => line.search(new RegExp(pattern)) === 0);
}

function pivotLine(lines: string[], matches: string[], cursorLine: number): number {
  let index = Math.min(cursorLine, lines.length - 1);
  while (index >= 0) {
    if (matchesAtStart(lines[index], matches)) {
      return index + 1;
    }
    index -= 1;
  }
  return 0;
}

export function computeLayout(
  text: string,
  screen: Screen,
  atlas: AtlasMetrics,
  matches: string[] = [],
  tags: string[] = [],
  cursorLine = 0,
): ClientLayoutResult {
  const lines = text.split("\n");
  const start = matches.length > 0 ? pivotLine(lines, matches, cursorLine) : 0;

  const glyphs: LayoutGlyph[] = [];
  let overflowHorizontal = false;
  let overflowVertical = false;
  let y = screen.y;

  for (let lineIndex = start; lineIndex < lines.length; lineIndex += 1) {
    if (y > SCREEN_HEIGHT) {
      overflowVertical = true;
      break;
    }

    // Limites (matches) sao estruturais: casam na linha crua, mesmo que as tags
    // as apaguem para exibicao.
    if (matches.length > 0 && matchesAtStart(lines[lineIndex], matches)) {
      break;
    }

    const stripped = stripTags(lines[lineIndex], tags);
    if (stripped.length === 0) {
      // linha sem glifos (delimitador/slot vazio): nao ocupa espaco vertical
      continue;
    }

    let x = screen.x;
    for (const character of stripped) {
      const codepoint = character.codePointAt(0) ?? 0;
      const sprite = atlas.glyphs[String(codepoint)];
      const advance = sprite ? sprite.advance : atlas.default_advance;
      const leading = sprite ? sprite.leading : 0;
      glyphs.push({
        codepoint,
        index: sprite ? sprite.index : -1,
        x: x + leading,
        y,
      });
      x += advance + 1;
      if (x > SCREEN_WIDTH) {
        overflowHorizontal = true;
        break;
      }
    }
    y += screen.new_line;
  }

  return { glyphs, overflowHorizontal, overflowVertical };
}
